### PR TITLE
Added Multi-Factor Authentication

### PR DESCRIPTION
**Pull request by chesspro13**
Added A tab in the options menu to enable Multi-Factor Authentication (MFA) in the form of Time-Based One Time Passwords (TOTP) for a layer of security.

White space corrected through the use of Prettier. Format standardization still needed issue #52 

Addresses feature request #68

### CODE REVIEW

**Comment by eliandoran** [comment]
@chesspro13 , since you're still working on the PR, I will put it in draft for now.
Feel free to put it into ready once you complete your development.
Keep up the good work!

**Comment by chesspro13** [comment]
TOTP functionality is complete. I'm going through and rewriting everything to try to match the original formatting as much as possible.  @eliandoran do you have a formatting config?

File: src/services/encryption/totp_secret.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
+'use strict';
```
Comment (chesspro13) [inline src/services/encryption/totp_secret.ts:1]: Remove password references

**Review by chesspro13 — COMMENTED** [review]
Changes required

File: src/public/app/widgets/type_widgets/options/multi_factor_authentication.js [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        }
+
+        server
+            .post('totp/set', {
+                secret: this.$totpSecretInput.val(),
+                password: this.$password.val()
+            })
+            .then((result) => {
+                if (result.success) {
+                    toastService.showError('Password has been changed. Trilium will be reloaded after you press OK.');
```
[the remark this answers is not in the document]
Comment (chesspro13) [reply src/public/app/widgets/type_widgets/options/multi_factor_authentication.js:198]: Changed

File: src/routes/api/totp.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import options = require('../../services/options');
```
[the remark this answers is not in the document]
Comment (chesspro13) [reply src/routes/api/totp.ts:1]: Removed unnecessary code

**Comment by chesspro13** [comment]
@eliandoran, just to clarify, the policy is to request review when merging into the develop branch, correct? Am I correct in assuming I can merge this since it's a pull request from my fork to this MFA branch?

**Comment by eliandoran** [comment]
@chesspro13 , generally the process is to open a PR that targets the main development branch (develop, in our case). The reason is that a PR is basically a way of asking to merge code into the main branch, with support for reviews.

Generally, you create a branch either on your fork on this repo (I prefer to create branches directly on the TriliumNext repo) and then create the PR. No need to have a branch both on the fork and on TriliumNext. 

In this case, my recommendation is to edit this PR (top-right) and change the base branch to develop.

**Comment by chesspro13** [comment]
Going to bundle this feature along with OAuth. Closing pull request.

**Pull request closed without merging**